cors.go

```diff
@@ -1,22 +1,113 @@
 package handlers
 
-import "net/http"
+import (
+	"net/http"
+	"strconv"
+	"strings"
+)
 
 // CORSOption represents a functional option for configuring the CORS middleware.
 type CORSOption func(*cors) error
 
 type cors struct {
-	h              http.Handler
-	allowedHeaders []string
-	allowedMethods []string
-	allowedOrigins []string
-	maxAge         int
-	ignoreOptions  bool
+	h                http.Handler
+	allowedHeaders   []string
+	allowedMethods   []string
+	allowedOrigins   []string
+	exposedHeaders   []string
+	maxAge           int
+	ignoreOptions    bool
+	allowCredentials bool
 }
 
+var (
+	defaultCorsMethods = []string{"GET", "HEAD", "POST"}
+	defaultCorsHeaders = []string{"Accept", "Accept-Language", "Content-Language"}
+	nilHandlerFunc     = http.HandlerFunc(nilHandler)
+)
+
+const (
+	corsOptionMethod           string = "OPTIONS"
+	corsAllowOriginHeader      string = "Access-Control-Allow-Origin"
+	corsExposeHeadersHeader    string = "Access-Control-Expose-Headers"
+	corsMaxAgeHeader           string = "Access-Control-Max-Age"
+	corsAllowMethodsHeader     string = "Access-Control-Allow-Methods"
+	corsAllowHeadersHeader     string = "Access-Control-Allow-Headers"
+	corsAllowCredentialsHeader string = "Access-Control-Allow-Credentials"
+	corsRequestMethodHeader    string = "Access-Control-Request-Method"
+	corsRequestHeadersHeader   string = "Access-Control-Request-Headers"
+	corsOriginHeader           string = "Origin"
+	corsVaryHeader             string = "Vary"
+	corsOriginMatchAll         string = "*"
+)
+
 func (ch *cors) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	handler := ch.h
+	defer func() {
+		handler.ServeHTTP(w, r)
+	}()
```

Comment:
I would probably remove the `defer` here and just call `ch.h.ServeHTTP(w, r)` as needed. There's only a couple of spots where you want to do that - `ignoreOptions` and on successfully passing the requirements. 

The handler will also be called after a `w.WriteHeader(errorStatus); return` block as well, which I don't think is what you want.